### DISCUSSION TITLE
Bambi's hierarchical modeling

### DISCUSSION THREAD

**Discussion by MatteoBenaud** [bambinos/bambi#808] [Q&A] [answered]
Hello everyone, I have a question regarding the theory behing bambi and specificly **hierarchical** modeling.

In traditional hierarchical Bayesian modeling, it is customary to incorporate hyper-priors to ensure that the resulting prior adequately captures the average effect across all categories.

However, upon reviewing the implementation in Bambi, for example [here](https://bambinos.github.io/bambi/notebooks/multi-level_regression.html), I noticed that there is only one prior specified for sigma. This raised a question regarding **the absence of a hyper-prior on mu**.

<img width="398" alt="image" src="https://github.com/bambinos/bambi/assets/73585675/b982af9b-a9e2-40a0-ad39-0fad616c57b3">

Could you help me elucidate the reason of this choice ? 
Is it possible to implement an hyperprior on mu ?

Thanks a lot !

**Comment by tomicapretto** [comment]
Hi @MatteoBenaud this is a great question and it's actually something I wanted to incorporate into our docs for a long time. I'll make an attempt at explaining how it works.

Bambi, as most (or all?) formula based interfaces, have some frequentist flavor when doing hierarchical (mixed, multi-level, etc.) modeling. In frequentist stats a mixed effects models for the mean (fixed and random effects) is written as follows

$$
\boldsymbol{\mu} = \boldsymbol{X}\boldsymbol{\beta} + \boldsymbol{Z}\boldsymbol{u}
$$

The first dot product represents the fixed effects and the second dot product represents the random effects. Bambi is always following this same structure. 

Let's present a simpler example. Let's imagine we have J groups and we want partially pooled intercepts. Using distributions we can write this model as 

$$
\begin{aligned}
Y_i      &\sim \text{Normal}(m_{j[i]}, s) \\
m_j &\sim \text{Normal}(\mu, \sigma) & \text{for all } j =  1 \dots J\\
\mu &\sim \text{Normal}(...) \\
\sigma &\sim \text{HalfNormal}(...) \\
s &\sim \text{Something}(...)
\end{aligned}
$$

So each group has its own mean $m_j$, and these means all come from the same normal hyperprior with mean $\mu$ and standard deviation $\sigma$. 

However, this is not the only way we can write this model. Have a look at the following:

$$
\begin{aligned}
Y_i    &\sim \text{Normal}(\mu + m'_{j[i]}, s) \\
m'_j &\sim \text{Normal}(0, \sigma) & \text{for all } j =  1 \dots J \\
\mu &\sim \text{Normal}(...) \\
\sigma &\sim \text{HalfNormal}(...) \\
s &\sim \text{Something}(...)
\end{aligned}
$$

Notice the $m'_j$ are now centered at 0.  These models are equivalent. However, the second one allows for a different "view". We can think of $\mu$ as a common mean and $m'_j$ as a group specific deflection around. This is _exactly_ what Bambi represents with `1` and `(1 | Pig)` in the model. `1` is the common intercept, and `(1 | Pig)` is the pig specific deviation around that common intercept. As in the math notation, the sum of these two components give the intercept for each pig. 

This is the reason why you see `(1 | Pig) ~ Normal(0, HalfNormal(...))`. The group-specific effect is a deviation around the common effect. The same idea applies to slopes. Ah, and `1` goes into the X * beta part, and `1 | Pig` goes into the Z * u part in the frequentist notation.

I don't want to extend too much here. I think the main idea is summarized above, but please feel free to ask for further clarification :)

**Reply by MatteoBenaud** [reply]
Hi @tomicapretto, 

Thanks a lot for your response. It's a lot more clear to me ! 

Just to confirm, this means that : 

```
Formula : Weight ~ Time + (Time|Pig)
Common-level effects
	Intercept ~ Normal(mu: 60, sigma: 133)
	Time ~ Normal(mu: 46, sigma: 18)
Group-level effects
	1|Pig ~ Normal(mu: 0, sigma: HalfNormal(sigma : 133))
	Time|Pig ~ Normal(mu : 0, sigma : HalfNormal(sigma: 18))
```

Is equivalent to : 

```
Formula : Weight ~ (Time|Pig)
Group-level effects
	1|Pig ~ Normal(mu: Normal(mu: 60, sigma: 133), sigma: HalfNormal(sigma : 133))
	Time|Pig ~ Normal(mu : Normal(mu: 46, sigma: 18), sigma : HalfNormal(sigma: 18))
```
?

If what I'm saying is true, do you know if even if from a theorical point of view it represents the same problem, it can imply differences on modelization side with bambi ? 
I'm thinking of MCMC sampling for example, I wonder if a 2 stage prior sampling on gruop level effects is the same as 2 single samplings for common level and group-level effects. 

Again, thanks a lot !!

**Reply by MatteoBenaud** [reply]
Hi @tomicapretto

Do you have an opinion on the question above? 
And if not do you know who I could contact to find out?

Thanks a lot again :)

**Reply by tomicapretto** [reply]
Hi @MatteoBenaud, sorry I never saw your message!

Yes! You're right about the equivalence of the two things. A small caveat in the second formula: it has an implicit intercept. So we should write `Weight ~ 0 + (1 + Time|Pig)` which makes is very explicit that there's no global intercept (the `0`) and there's a partially pooled intercept for each Pig (the `1`).

However, I don't understand the following very well:

> do you know if even if from a theorical point of view it represents the same problem, it can imply differences on modelization side with bambi ?

Yes it represents the same model, the parametrization is different and it can lead to different performances when using NUTS to draw samples, but I don't understand the part about "differences on modelization side with bambi".

As for the following

>I'm thinking of MCMC sampling for example, I wonder if a 2 stage prior sampling on gruop level effects is the same as 2 single samplings for common level and group-level effects.

I'm not familiar with "2 stage prior sampling" and "2 single samplings". Sorry! (maybe if you are able to explain it I can add my two cents)

**Reply by MatteoBenaud** [reply]
Hi @tomicapretto 
Thanks a lot for your answer ! 

Maybe my questions were not very clear because of a misunderstanding on how sampling works with hyper priors. 

`Time ~ Normal(mu: 46, sigma: 18)`
I imagine the sampling here to pick a value inside this distribution, accept it/ reject it according to the posterior metric.
Each samples accepted or rejected leads to a chain representing the posterior probability. 

`Time|Pig ~ Normal(mu : 0, sigma : 18)`
For this one I imagine it is the same, but a different value is sampled for each pig. 

`Time|Pig ~ Normal(mu : Normal(mu: 46, sigma: 18), sigma : HalfNormal(sigma: 18))`
This one being hierarchical, I imagine a first sampling from Normal(mu:46, sigma:18) and HalfNormal(sigma: 18) and then the final sampling value for each pig from the Normal law with the mu and sigma picked from hyperpriors. 
That is why I called it a 2 stage prior sampling. 

Do not hesitate to tell me if I'm wrong !

**Reply by MatteoBenaud** [reply]
Additional question : 

Maybe it is not the right place, I can move in to a new discussion. 

I tried to compute some models with hyper priors. Doing it this way : 
`Formula : Weight ~ (Time|Pig)
Group-level effects
	1|Pig ~ Normal(mu: Normal(mu: 60, sigma: 133), sigma: HalfNormal(sigma : 133))
	Time|Pig ~ Normal(mu : Normal(mu: 46, sigma: 18), sigma : HalfNormal(sigma: 18))`
So not like the bambi's tutorial way. 

It worked fine until I added Truncated Normal priors. 
I got very poor results with only one value sampled : 
![image](https://github.com/user-attachments/assets/c981c1c9-c319-4fa4-8ba7-a2ee9324b1fc)

From my understanding, for some variables the sampler is starting outside the TruncatedNormal distribution (where the probability is 0). And I think it does not find the way to move to the right place to sample as the grandient is 0 too and remains stuck at the starting point.

Do you know if there is a way to force the starting point to start exactly in the middle of my prior ? 
I use nuts_numpyro as inference method. 

I tried the different init method but None worked. 

Thanks again for your help :)

**Reply by tomicapretto** [reply]
> Hi @tomicapretto Thanks a lot for your answer !
> 
> Maybe my questions were not very clear because of a misunderstanding on how sampling works with hyper priors.
> 
> `Time ~ Normal(mu: 46, sigma: 18)` I imagine the sampling here to pick a value inside this distribution, accept it/ reject it according to the posterior metric. Each samples accepted or rejected leads to a chain representing the posterior probability.
> 
> `Time|Pig ~ Normal(mu : 0, sigma : 18)` For this one I imagine it is the same, but a different value is sampled for each pig.
> 
> `Time|Pig ~ Normal(mu : Normal(mu: 46, sigma: 18), sigma : HalfNormal(sigma: 18))` This one being hierarchical, I imagine a first sampling from Normal(mu:46, sigma:18) and HalfNormal(sigma: 18) and then the final sampling value for each pig from the Normal law with the mu and sigma picked from hyperpriors. That is why I called it a 2 stage prior sampling.
> 
> Do not hesitate to tell me if I'm wrong !

I think I understand what you have in mind. But that's not how posterior sampling works.

In Bambi (because of PyMC), we are able to create an unnormalized log-posterior density function. What is that?

Bayes rule is:

$$
p(\theta \mid y) = \frac{p(y \mid \theta) p(\theta)}{\int p(y \mid \theta) p(\theta) d\theta}
$$

The $p(\theta \mid y)$ is known as the posterior density function.

The denominator can be seen as a normalizing constant. In that case, we usually write

$$
p(\theta \mid y) \propto p(y \mid \theta) p(\theta)
$$

So we _know_ the posterior density up to a normalizing constant. There are methods that allow us to extract samples from an arbitrary distribution if we known the density up to a normalizing constant, that's great. One of those methods is Metropolis-Hastings.

But Metropolis-Hastings can't handle the complexity of the posteriors have in the models we usually build. For that reason, we need more advanced methods. One of those is Hamiltonian Monte Carlo (think of NUTS as an advanced dynamic Hamiltonian Monte Carlo).
For reasons I'm not covering here, HMC works with the logarithm of the unnormalized pdf. That is:


$$
\log(p(\theta \mid y)) \propto \log(p(y \mid \theta)) + \log(p(\theta))
$$

In other words, while we need an initial draw from where to start (which _could_ be taken from the prior), the priors are not used to generate draws when we get draws from the posterior. The priors affect through their contribution to the logp, which otherwise would be just the likelihood.

---

If you want to see how HMC works, I have a (not extremely polished) Shiny app written in R.

https://github.com/tomicapretto/shiny-hmc

**Reply by tomicapretto** [reply]
@MatteoBenaud an extra note. The process you describe would be correct if the goal is to sample from the prior, without the accept/reject metric.

**Reply by tomicapretto** [reply]
> Additional question :
> 
> Maybe it is not the right place, I can move in to a new discussion.
> 
> I tried to compute some models with hyper priors. Doing it this way : `Formula : Weight ~ (Time|Pig) Group-level effects 1|Pig ~ Normal(mu: Normal(mu: 60, sigma: 133), sigma: HalfNormal(sigma : 133)) Time|Pig ~ Normal(mu : Normal(mu: 46, sigma: 18), sigma : HalfNormal(sigma: 18))` So not like the bambi's tutorial way.
> 
> It worked fine until I added Truncated Normal priors. I got very poor results with only one value sampled : ![image](https://private-user-images.githubusercontent.com/73585675/351372438-c981c1c9-c319-4fa4-8ba7-a2ee9324b1fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxNDYxNDksIm5iZiI6MTcyMzE0NTg0OSwicGF0aCI6Ii83MzU4NTY3NS8zNTEzNzI0MzgtYzk4MWMxYzktYzMxOS00ZmE0LThiYTctYTJlZTkzMjRiMWZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA4VDE5MzcyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkOTEzNjZlYWRiMTIxZTljZDVhYzA0YmM5YWZmZmQzMTI4ZmRkOTE0NmI0MzIzMGUxYjdmNzczMTU1MTU4NTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Bq4YqYfCr3I4cGqd4o0zQp425eOmnK_h7_reUpoltJ4)
> 
> From my understanding, for some variables the sampler is starting outside the TruncatedNormal distribution (where the probability is 0). And I think it does not find the way to move to the right place to sample as the grandient is 0 too and remains stuck at the starting point.
> 
> Do you know if there is a way to force the starting point to start exactly in the middle of my prior ? I use nuts_numpyro as inference method.
> 
> I tried the different init method but None worked.
> 
> Thanks again for your help :)

It would be good if you could make it a separate discussion, and even better if you could provide the code you used. I would be glad to help :)

**Reply by peeeffchang** [reply]
A related question: can Bambi handle multi-level DAG? For example, I would like to build a model with a DAG that looks like

A->B->D<-C<-A

But it's not clear whether/how it can be specified in Bambi.

**Reply by tomicapretto** [reply]
Bambi works with parametric models. It's certainly possible in those cases where the outcome distribution has at least two parameters. For example,

$$
\begin{aligned}
\mu_i &= \alpha_0 + \alpha_1 \text{Age}_i \\
\sigma_i &= \exp(\beta_0 + \beta_1 \text{Age}_i) \\
Y_i \mid \mu_i, \sigma_i &\sim \text{Normal}(\mu_i, \sigma_i^2)
\end{aligned}
$$

omitting priors for $\alpha$ and $\beta$ parameters. In that notation, D is the outcome, B is the mean mu, C is the standard deviation, and A is age.

In Bambi that would be

```python
formula = bmb.Formula(
    "y ~ 1 + age", # mu is implicit here 
    "sigma ~ 1 + age",
)
model = bmb.Model(formula, data)
```

where I'm being explicit about the inclusion of the intercept in both cases.

**Reply by peeeffchang** [reply]
I see, thank you very useful info. But does that also mean something like a set of hierarchical SCMs cannot be handled in Bambi?

**Reply by tomicapretto** [reply]
Do you have any example of it implemented in PyMC or in other language? I'm sorry but I'm not very familiar with SCMs

**Reply by peeeffchang** [reply]
One example would be like this:
[https://www.pymc-labs.com/blog-posts/causal-analysis-with-pymc-answering-what-if-with-the-new-do-operator/](https://www.pymc-labs.com/blog-posts/causal-analysis-with-pymc-answering-what-if-with-the-new-do-operator/)

**Reply by tomicapretto** [reply]
This one? 

```python
with pm.Model(coords_mutable={"i": [0]}) as model_generative:
    # priors on Y <- C -> Z
    beta_y0 = pm.Normal("beta_y0")
    beta_cy = pm.Normal("beta_cy")
    beta_cz = pm.Normal("beta_cz")
    # priors on Z -> Y causal path
    beta_z0 = pm.Normal("beta_z0")
    beta_zy = pm.Normal("beta_zy")
    # observation noise on Y
    sigma_y = pm.HalfNormal("sigma_y")
    # core nodes and causal relationships
    c = pm.Normal("c", mu=0, sigma=1, dims="i")
    z = pm.Bernoulli("z", p=pm.invlogit(beta_z0 + beta_cz * c), dims="i")
    y_mu = pm.Deterministic("y_mu", beta_y0 + (beta_zy * z) + (beta_cy * c), dims="i")
    y = pm.Normal("y", mu=y_mu, sigma=sigma_y, dims="i")

pm.model_to_graphviz(model_generative)
```

**Reply by peeeffchang** [reply]
Yep that's the one describing the system

**Reply by tomicapretto** [reply, marked as answer]
I'm afraid it's not possible to implement that model in Bambi because of:

* The multiple likelihoods (`c`, `z`, and `y`)
* The dependency between them (e.g. `z` and `c` go into `y_mu`).

It's possible one could write a custom family that hacks through Bambi internals and implements this PyMC model. But if one is willing to do such amount of work, it's just simpler to use plain PyMC.

**Reply by peeeffchang** [reply]
@tomicapretto Noted thanks
